src/tunnel.cc

```diff
@@ -894,6 +1021,7 @@ TunnelStateData::noteConnection(HappyConnOpener::Answer &answer)
         syncHierNote(answer.conn, request->url.host());
         saveError(error);
         answer.error.clear(); // savedError has it now
+        // XXX: should honor clientExpectsConnectResponse() before replying. Also check other occurrences.
         sendError(savedError, "tried all destinations");
```

Comment:
Please enact this XXX before merging. It looks like the correct actions will be performed by:
```suggestion
        retriable = false;
        retryOrBail("tried all destinations");
```

Reply:
You are right that we could easily fix this specific case but, unfortunately, these 'other occurrences' do exist (see, e.g., TunnelStateData::noteDestinationsEnd()). So, we would have to move then this XXX there but the same question arises again. I think it would be better to resolve all these cases in a separate dedicated PR.

Reply:
> it would be better to resolve all these cases in a separate dedicated PR.

@eduard-bagdasaryan, agreed. This fix is not directly related to this PR and requires non-trivial likely-to-cascade changes. Since those changes affect traffic on the wire, a dedicated description would help admins to asses whether the fix is related to problems they suffer from. The fix deserves its own PR.

In general, we should avoid penalizing folks for reporting bugs. If a developer adds an XXX to some old code, and that XXX _can be_ interpreted as lying outside the PR scope, we should not press the developer into addressing that XXX. Otherwise, developers will simply stop adding valuable XXXs and TODOs, hurting everybody!

Reply:
Sending a plain-text error page to a client which is not speaking plain-text HTTP. The client may be speaking some unknown protocol which can be corrupted in unpredictable ways by the error page response.

We should not be accepting a PR which actively *adds* cases of traffic corruption. Especially when that protection already existed with the previous sendError() call and is being deleted by this PR. It should have been moved along with the sendError() call.

Reply:
> Especially when that protection already existed with the previous sendError() 

This is incorrect: the old sendError() lacked such protection.

Reply:
Sorry for not being clear, I mean the code sequence in tunnelEstablishmentDone() protecting sendError() calls:
```
 if (!clientExpectsConnectResponse()) {
        // closing the non-HTTP client connection is the best we can do
        debugs(50, 3, server.conn << " closing on CONNECT-to-peer error");
        server.closeIfOpen();
        return;
    }
...
    sendError(error, "tunneler returns error");
```

The if-condition preventing error delivery is gone in PR code paths. This XXX is about the new PR code re-opening old bugs by removal of that condition.

IMO it is reasonable to expect this PR to avoid adding at least the easily solved occurrences (like this marked one) to minimize new bugs. Ones that are found to be not so simple can be deferred with an XXX as Alex says - at least then "Also check other occurrences" part is done and all remaining bugs marked so nobody mistakenly copies the bad code around before that deferred fix is done.

Reply:
> The if-condition preventing error delivery is gone in PR code paths. 

If your concern is only about tunnelEstablishmentDone() - I see no problem there, because retryOrBail() takes care of not calling sendError() if clientExpectsConnectResponse() returns false.

Reply:
That is *one* path where the now-deleted condition code enters retryOrBail() and reaches sendError(). There are many paths through retryOrBail() which do not reach that safe sendError() - as intended most of those exit and loop back into logic where other errors may need sending. eg. follow a winding path around to hit this XXX marked sendError() or any of the "other occurrences" alluded to.
 Thus; all those logic paths have this newly added "regression" bug.

So ... what about all the other sendError() occurances? please do at least locate and XXX marking part so the "Also check other occurrences" text can be erased. If possible fix so they do not need that period of annoying XXX behaviour.

Reply:
Sigh. This discussion and associated waste and delays is exactly why every time I want to document an accidentally discovered out-of-scope problem, I tell myself "Stop, Amos will make you regret this good deed!". I am a fool for hoping that "the next time it will be different". Again and again, we are being penalized for discovering and documenting out-of-scope problems.

Reply:
FTR, Factory will address this XXX. Since the XXX is unrelated to this PR, there will be a separate PR addressing it.